include/zephyr/sensing/sensing_sensor.h

```diff
@@ -0,0 +1,594 @@
+ *
+ * @param index The value fields index to be set, -1 for all fields (global).
+ *
+ * @param value The data buffer to receive value.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+typedef int (*sensing_sensor_get_range_t)(
+		const struct device *dev,
+		int index, uint32_t *value);
+
+/**
+ * @brief Set current sensor's hardware fifo size
+ *
+ * Some sensors especially for physical sensors, support hardware fifo, this API can
+ * configure the current fifo size.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param samples The sample number to set for fifo.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+typedef int (*sensing_sensor_set_fifo_t)(
+		const struct device *dev,
+		uint32_t samples);
+
+/**
+ * @brief Get current sensor's hardware fifo size
+ *
+ * Some sensors especially for physical sensors, support fifo, this API can
+ * get the current fifo size.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param samples The data buffer to receive the fifo sample number.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+typedef int (*sensing_sensor_get_fifo_t)(
+		const struct device *dev,
+		uint32_t *samples);
+
+/**
+ * @brief Set current sensor data offset
+ *
+ * Some sensors especially for physical sensors, such as accelerometer senors,
+ * as data drift, need configure offset calibration.
+ *
+ * Since each sensor type may have multiple data fields for it's value, this
+ * API support set separated offset for each data field, or global offset for
+ * all data fields.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param index The value fields index to be set, -1 for all fields (global).
+ *
+ * @param value The offset value to be set.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+typedef int (*sensing_sensor_set_offset_t)(
+		const struct device *dev,
+		int index, int32_t value);
+
+/**
+ * @brief Get current sensor data offset
+ *
+ * Some sensors especially for physical sensors, such as accelerometer senors,
+ * as data drift, need configure offset calibration.
+ *
+ * Since each sensor type may have multiple data fields for it's value, this
+ * API support get separated offset for each data field, or global offset for
+ * all data fields.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param index The value fields index to be set, -1 for all fields (global).
+ *
+ * @param value The data buffer to receive the offset value.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+typedef int (*sensing_sensor_get_offset_t)(
+		const struct device *dev,
+		int index, int32_t *value);
+/**
+ * @struct sensing_sensor_api
+ * @brief Sensor callback api
+ *
+ * A sensor must register this callback API during sensor registration.
+ */
+struct sensing_sensor_api {
```

Comment:
Please remove this struct from this PR. I still don't see any use for it and I'm not seeing anything in the sample that will convince me otherwise.

- Getting data from the subsystem should be done by opening a connection to the subsystem's desired sensor and using the callback mechanism like every other client.
- To enable these get/set function, simply use the existing These sensors can simply be clients of the subsystem and use `sensing_open_sensor()` with their own callbacks. The key is that `SENSOR_DEVICE_DT_DEFINE` should _NOT_ be used or it'll confuse the driver utilities that will think it's a normal Kernel level device.

Reply:
This `sensing_sensor_api` callbacks is for sensor to response `Sensing Subsystem`, sensor which implemented these callbacks will be management and scheduled by `Sensing Subsystem`. 

Below is the sensor configuration flow: 
![sensor_config_flow](https://github.com/zephyrproject-rtos/zephyr/assets/45843554/4875e813-88d2-4c37-842e-9dbb95eadd71)

As you can see,  when App call `sensing_set_config(handle_hinge,...)` to config report interval to the `hinge angle` virtual sensor, the `hinge angle` virtual sensor's `set_interval()` callback will be called, in this callback, the `hinge angle` virtual sensor can call the unified `sensing_set_config(handle_base_acc,...)` and `sensing_set_config(handle_lid_acc,...)` to config it's reporter's report interval. 

As you see, we used same `handle` concepts and API (`sensing_set_config()`) for both App client and Sensor client to config it's reporters.  

And below is the sensor data flow: 
![sensor_data_flow](https://github.com/zephyrproject-rtos/zephyr/assets/45843554/ff6568fb-c285-421d-95ff-ce967b59f471)

In this flow, we can see the `hinge angle` virtual sensor can implement the `process()` callback, in this callback, `hinge angle` receives reporter sensors' data, doing hinge angle computing and return to `Sensing Subsystem`, at last `Sensing Subsystem` return to App through `data_event` callback.  

As you said, ether Application client or sensor client consumers, they do have similar config/data connections with reporter sensors, that's why we used the unified `handler` for these connections, and internally they do have same path. 

The major differences is the `handler` for Application(s) need the Application(s) to explicitly call `sensing_sensor_open()` to based on usages, and the `handler`s between sensors will be auto generated by `Sensing Subsystem` based on devicetree configuration for the sensors reporting topology (the bluebox in below diagram), since this is all determined at build time base on devicetree, so the `Sensing Subsystem` can auto open reporter sensors with `handler`s for their client sensors, Application(s) not need care bout this, and just open the desired sensor(s), make sense ? 

![sensor_top](https://github.com/zephyrproject-rtos/zephyr/assets/45843554/76ef640a-ae90-4101-afd2-af99f63da691)

The `sensing_sensor_api` defines a standard callback APIs for sensors implementation, it can make sensor developers just focus on the callback APIs implementation, and make all the sensors which followed these callback APIs work together and reusable cross platforms and projects. Make sense ?  

In short, Application(s) use APIs in `sensing.h` to access any sensor(s), vendors use `sensing_sensor_api` in `sensing_sensor.h` to define and implement sensors, others all handled by the `Sensing Subsystem` (scheduling, multiple clients, arbitration, sensor reporting topology ...). This is the core concepts of the `Sensing Subsystem` introduced in this PR, and can help build unified sensing ecosystems, although in this PR we haven't implemented a real sensor yet (will be in follow PRs), we still want keep this core concepts complete in this PR.